<h1 align="center"> London vs New York</h1>
<h2 align="center">Clustering the Neighbourhoods of London and New York</h2>

<p align ="center"> Vyas Cholayil
<br>
<br>
7th Nov 2020
</p>




# Introduction

London and New York are two of my favorite cities in this world. These cities were both happening then and now. A lot has changed over the years and we now take a look at how the cities have grown. 

London and New york are quite the popular tourist and vacation destinations for people all around the world. They are diverse and multicultural and offer a wide variety of experiences that is widely sought after. We try to group the neighbourhoods of London and New York respectively and draw insights to what they look like now.

# Business Problem


The aim is to help tourists choose their destinations depending on the experiences that the neighbourhoods have to offer and what they would want to have. This also helps people make decisions if they are thinking about migrating to London or Paris or even if they want to relocate neighbourhoods within the city. Our findings will help stakeholders make informed decisions and address any concerns they have including the different kinds of cuisines, provision stores and what the city has to offer. 



# Data Description

We require geolocation data for both London and Paris. Postal codes in each city serve as a starting point. Using Postal codes we use can find out the neighbourhoods, boroughs, venues and their most popular venue categories.


## London

To derive our solution, We scrape our data from https://en.wikipedia.org/wiki/List_of_areas_of_London

This wikipedia page has information about all the neighbourhoods, we limit it London.

1. *borough* : Name of Neighborhood
2. *town* : Name of borough
3. *post_code* : Postal codes for London.

This wikipedia page lacks information about the geographical locations. To solve this problem we use ArcGIS API

### ArcGIS API

ArcGIS Online enables you to connect people, locations, and data using interactive maps. Work with smart, data-driven styles and intuitive analysis tools that deliver location intelligence. Share your insights with the world or specific groups. 

More specifically, we use ArcGIS to get the geo locations of the neighbourhoods of London. The following columns are added to our initial dataset which prepares our data. 

4. *latitude* : Latitude for Neighborhood
5. *longitude* : Longitude for Neighborhood

## New York

To derive our solution, We leverage JSON data available at  https://geo.nyu.edu/catalog/nyu_2451_34572

The JSON file has data about all the neighbourhoods in France, we limit it to Paris.

1. *borough* : Name of borough
2. *neighborhood* : Name of Neighborhoods in New York
3. *latitude* : Latitude for Neighborhood
4. *longitude* : Longitude for Neighborhood

## Foursquare API Data

We will need data about different venues in different neighbourhoods of that specific borough. In order to gain that information we will use "Foursquare" locational information. Foursquare is a location data provider with information about all manner of venues and events within an area of interest. Such information includes venue names, locations, menus and even photos. As such, the foursquare location platform will be used as the sole data source since all the stated required information can be obtained through the API.

After finding the list of neighbourhoods, we then connect to the Foursquare API to gather information about venues inside each and every neighbourhood. For each neighbourhood, we have chosen the radius to be 500 meters.

The data retrieved from Foursquare contained information of venues within a specified distance of the longitude and latitude of the postcodes. The information obtained per venue as follows:

1. *Neighbourhood* : Name of the Neighbourhood
2. *Neighbourhood Latitude* : Latitude of the Neighbourhood
3. *Neighbourhood Longitude* : Longitude of the Neighbourhood
4. *Venue* : Name of the Venue
5. *Venue Latitude* : Latitude of Venue
6. *Venue Longitude* : Longitude of Venue
7. *Venue Category* : Category of Venue


Based on all the information collected for both London and Paris, we have sufficient data to build our model. We cluster the neighbourhoods together based on similar venue categories. We then present our observations and findings. Using this data, our stakeholders can take the necessary decision.

# Methodology

We will be creating our model with the help of Python so we start off by importing all the required packages.

In [2]:
!pip install folium

import pandas as pd
import requests
import numpy as np
import matplotlib.cm as cm
import matplotlib.colors as colors
import folium

# import k-means for the clustering stage
from sklearn.cluster import KMeans

     |████████████████████████████████| 93 kB 3.2 MB/s eta 0:00:011


The approach taken here is to explore each of the cities individually, plot the map to show the neighbourhoods being considered and then build our model by clustering all of the similar neighbourhoods together and finally plot the new map with the clustered neighbourhoods. We draw insights and then compare and discuss our findings.

# Exploring London

### Neighbourhoods of London

We begin to start collecting and refining the data needed for the our business solution to work.

### Data Collection

To get the neighbourhoods in london, we start by scraping the list of areas of london wiki page.

In [3]:
url_london = "https://en.wikipedia.org/wiki/List_of_areas_of_London"
wiki_london_url = requests.get(url_london)
wiki_london_url

<Response [200]>

Response 200 means that we are able to make the connection

In [4]:
wiki_london_data = pd.read_html(wiki_london_url.text)
wiki_london_data

[                                                   0
 0  Map all coordinates in "Category:Areas of Lond...
 1                 Download coordinates as: KML · GPX,
             Location                     London borough       Post town  \
 0         Abbey Wood              Bexley, Greenwich [7]          LONDON   
 1              Acton  Ealing, Hammersmith and Fulham[8]          LONDON   
 2          Addington                         Croydon[8]         CROYDON   
 3         Addiscombe                         Croydon[8]         CROYDON   
 4        Albany Park                             Bexley  BEXLEY, SIDCUP   
 ..               ...                                ...             ...   
 527         Woolwich                          Greenwich          LONDON   
 528   Worcester Park       Sutton, Kingston upon Thames  WORCESTER PARK   
 529  Wormwood Scrubs             Hammersmith and Fulham          LONDON   
 530          Yeading                         Hillingdon           HAYES   
 

Scraping the webpage gives us all the tables present on the page. We need the 2nd table, so selecting the 2nd table.

In [5]:
wiki_london_data = wiki_london_data[1]
wiki_london_data

Location                     London borough       Post town  \
0         Abbey Wood              Bexley, Greenwich [7]          LONDON   
1              Acton  Ealing, Hammersmith and Fulham[8]          LONDON   
2          Addington                         Croydon[8]         CROYDON   
3         Addiscombe                         Croydon[8]         CROYDON   
4        Albany Park                             Bexley  BEXLEY, SIDCUP   
..               ...                                ...             ...   
527         Woolwich                          Greenwich          LONDON   
528   Worcester Park       Sutton, Kingston upon Thames  WORCESTER PARK   
529  Wormwood Scrubs             Hammersmith and Fulham          LONDON   
530          Yeading                         Hillingdon           HAYES   
531         Yiewsley                         Hillingdon    WEST DRAYTON   

    Postcode district Dial code OS grid ref  
0                 SE2       020    TQ465785  
1              W3, W4       020    TQ205805  
2                 CR0       020    TQ375645  
3                 CR0       020    TQ345665  
4           DA5, DA14       020    TQ478728  
..                ...       ...         ...  
527              SE18       020    TQ435795  
528               KT4       020    TQ225655  
529               W12       020    TQ225815  
530               UB4       020    TQ115825  
531               UB7       020    TQ063804  

[532 rows x 6 columns]

### Data Preprocessing

we remove the spaces in the column titles and then we add _ between words.

In [6]:
wiki_london_data.rename(columns=lambda x: x.strip().replace(" ", "_"), inplace=True)
wiki_london_data

Location                     London borough       Post_town  \
0         Abbey Wood              Bexley, Greenwich [7]          LONDON   
1              Acton  Ealing, Hammersmith and Fulham[8]          LONDON   
2          Addington                         Croydon[8]         CROYDON   
3         Addiscombe                         Croydon[8]         CROYDON   
4        Albany Park                             Bexley  BEXLEY, SIDCUP   
..               ...                                ...             ...   
527         Woolwich                          Greenwich          LONDON   
528   Worcester Park       Sutton, Kingston upon Thames  WORCESTER PARK   
529  Wormwood Scrubs             Hammersmith and Fulham          LONDON   
530          Yeading                         Hillingdon           HAYES   
531         Yiewsley                         Hillingdon    WEST DRAYTON   

    Postcode district Dial code OS_grid_ref  
0                 SE2       020    TQ465785  
1              W3, W4       020    TQ205805  
2                 CR0       020    TQ375645  
3                 CR0       020    TQ345665  
4           DA5, DA14       020    TQ478728  
..                ...       ...         ...  
527              SE18       020    TQ435795  
528               KT4       020    TQ225655  
529               W12       020    TQ225815  
530               UB4       020    TQ115825  
531               UB7       020    TQ063804  

[532 rows x 6 columns]

We see that few columns have no '_' between the words despite applying our function meaning that there are special characters

### Feature Selection

We need only the boroughs, Postal codes, Post town for further steps. We can drop the locations, dial codes and OS grid.

In [7]:
df1 = wiki_london_data.drop( [ wiki_london_data.columns[0], wiki_london_data.columns[4], wiki_london_data.columns[5] ], axis=1)

In [8]:
df1.head()

London borough       Post_town Postcode district
0              Bexley, Greenwich [7]          LONDON               SE2
1  Ealing, Hammersmith and Fulham[8]          LONDON            W3, W4
2                         Croydon[8]         CROYDON               CR0
3                         Croydon[8]         CROYDON               CR0
4                             Bexley  BEXLEY, SIDCUP         DA5, DA14

let's rename the Postcode district column and the london borough to something simpler

In [9]:
df1.columns = ['borough','town','post_code']
df1

borough            town  post_code
0                Bexley, Greenwich [7]          LONDON        SE2
1    Ealing, Hammersmith and Fulham[8]          LONDON     W3, W4
2                           Croydon[8]         CROYDON        CR0
3                           Croydon[8]         CROYDON        CR0
4                               Bexley  BEXLEY, SIDCUP  DA5, DA14
..                                 ...             ...        ...
527                          Greenwich          LONDON       SE18
528       Sutton, Kingston upon Thames  WORCESTER PARK        KT4
529             Hammersmith and Fulham          LONDON        W12
530                         Hillingdon           HAYES        UB4
531                         Hillingdon    WEST DRAYTON        UB7

[532 rows x 3 columns]

Let's remove the Square brackets [ ] and numbers from the borough column

In [10]:
df1['borough'] = df1['borough'].map(lambda x: x.rstrip(']').rstrip('0123456789').rstrip('['))
df1

borough            town  post_code
0                Bexley, Greenwich           LONDON        SE2
1    Ealing, Hammersmith and Fulham          LONDON     W3, W4
2                           Croydon         CROYDON        CR0
3                           Croydon         CROYDON        CR0
4                            Bexley  BEXLEY, SIDCUP  DA5, DA14
..                              ...             ...        ...
527                       Greenwich          LONDON       SE18
528    Sutton, Kingston upon Thames  WORCESTER PARK        KT4
529          Hammersmith and Fulham          LONDON        W12
530                      Hillingdon           HAYES        UB4
531                      Hillingdon    WEST DRAYTON        UB7

[532 rows x 3 columns]

Take the dimension of the dataframe

In [11]:
df1.shape

(532, 3)

We currently have 533 records and 3 columns of our data. It's time to perform Feature Engineering

### Feature Engineering

We can only focusing on the neighbourhoods of London, so performing the changes

In [12]:
df1 = df1[df1['town'].str.contains('LONDON')]
df1

borough                    town post_code
0                Bexley, Greenwich                   LONDON       SE2
1    Ealing, Hammersmith and Fulham                  LONDON    W3, W4
6                              City                  LONDON       EC3
7                       Westminster                  LONDON       WC2
9                           Bromley                  LONDON      SE20
..                              ...                     ...       ...
522                       Redbridge                  LONDON  IG8, E18
523       Redbridge, Waltham Forest  LONDON, WOODFORD GREEN       IG8
526                          Barnet                  LONDON       N12
527                       Greenwich                  LONDON      SE18
529          Hammersmith and Fulham                  LONDON       W12

[309 rows x 3 columns]

In [13]:
df1.shape

(309, 3)

We now have only 310 rows. We can proceed with our further steps. Getting some descriptive statistics

In [14]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 309 entries, 0 to 529
Data columns (total 3 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   borough    309 non-null    object
 1   town       309 non-null    object
 2   post_code  309 non-null    object
dtypes: object(3)
memory usage: 9.7+ KB


## Geolocations of the London Neighbourhoods

### ArcGis API

We need to get the geographical co-ordinates for the neighbourhoods to plot out map. We will use the arcgis package to do so. 

Arcgis doesn't have a limitation on the number of API calls made so it fits our use case perfectly.

In [15]:
pip install arcgis

Note: you may need to restart the kernel to use updated packages.


In [16]:
from arcgis.geocoding import geocode
from arcgis.gis import GIS
gis = GIS()

/opt/conda/envs/Python-3.7-main/lib/python3.7/site-packages/arcgis/features/_data/geodataset/utils.py:16: FutureWarning: The pandas.datetime class is deprecated and will be removed from pandas in a future version. Import from datetime module instead.
  pd.datetime,


Defining London arcgis geocode function to return latitude and longitude

In [17]:
def get_x_y_uk(address1):
   lat_coords = 0
   lng_coords = 0
   g = geocode(address='{}, London, England, GBR'.format(address1))[0]
   lng_coords = g['location']['x']
   lat_coords = g['location']['y']
   return str(lat_coords) +","+ str(lng_coords)

Checking sample data

In [18]:
c = get_x_y_uk('SE2')

In [19]:
c

'51.492450000000076,0.12127000000003818'

Looks good, We Copy over the postal codes of london to pass it into the geolocator function that we just defined above

In [20]:
geo_coordinates_uk = df1['post_code']    
geo_coordinates_uk

0           SE2
1        W3, W4
6           EC3
7           WC2
9          SE20
         ...   
522    IG8, E18
523         IG8
526         N12
527        SE18
529         W12
Name: post_code, Length: 309, dtype: object

Passing postal codes of london to get the geographical co-ordinates

In [21]:
coordinates_latlng_uk = geo_coordinates_uk.apply(lambda x: get_x_y_uk(x))
coordinates_latlng_uk

0       51.492450000000076,0.12127000000003818
1        51.51324000000005,-0.2674599999999714
6       51.51200000000006,-0.08057999999994081
7       51.51651000000004,-0.11967999999995982
9       51.41009000000008,-0.05682999999993399
                        ...                   
522    51.589770000000044,0.030520000000024083
523      51.50642000000005,-0.1272099999999341
526     51.615920000000074,-0.1767399999999384
527      51.48207000000008,0.07143000000002075
529      51.50645000000003,-0.2369099999999662
Name: post_code, Length: 309, dtype: object

### Latitude

Extracting the latitude from our previously collected coordinates

In [22]:
lat_uk = coordinates_latlng_uk.apply(lambda x: x.split(',')[0])
lat_uk

0      51.492450000000076
1       51.51324000000005
6       51.51200000000006
7       51.51651000000004
9       51.41009000000008
              ...        
522    51.589770000000044
523     51.50642000000005
526    51.615920000000074
527     51.48207000000008
529     51.50645000000003
Name: post_code, Length: 309, dtype: object

### Longitude

Extracting the Longitude from our previously collected coordinates

In [23]:
lng_uk = coordinates_latlng_uk.apply(lambda x: x.split(',')[1])
lng_uk

0       0.12127000000003818
1       -0.2674599999999714
6      -0.08057999999994081
7      -0.11967999999995982
9      -0.05682999999993399
               ...         
522    0.030520000000024083
523     -0.1272099999999341
526     -0.1767399999999384
527     0.07143000000002075
529     -0.2369099999999662
Name: post_code, Length: 309, dtype: object

We now have the geographical co-ordinates of the London Neighbourhoods.

We proceed with Merging our source data with the geographical co-ordinates to make our dataset ready for the next stage

In [24]:
london_merged = pd.concat([df1,lat_uk.astype(float), lng_uk.astype(float)], axis=1)
london_merged.columns= ['borough','town','post_code','latitude','longitude']
london_merged

borough                    town post_code  \
0                Bexley, Greenwich                   LONDON       SE2   
1    Ealing, Hammersmith and Fulham                  LONDON    W3, W4   
6                              City                  LONDON       EC3   
7                       Westminster                  LONDON       WC2   
9                           Bromley                  LONDON      SE20   
..                              ...                     ...       ...   
522                       Redbridge                  LONDON  IG8, E18   
523       Redbridge, Waltham Forest  LONDON, WOODFORD GREEN       IG8   
526                          Barnet                  LONDON       N12   
527                       Greenwich                  LONDON      SE18   
529          Hammersmith and Fulham                  LONDON       W12   

     latitude  longitude  
0    51.49245    0.12127  
1    51.51324   -0.26746  
6    51.51200   -0.08058  
7    51.51651   -0.11968  
9    51.41009   -0.05683  
..        ...        ...  
522  51.58977    0.03052  
523  51.50642   -0.12721  
526  51.61592   -0.17674  
527  51.48207    0.07143  
529  51.50645   -0.23691  

[309 rows x 5 columns]

In [25]:
london_merged.dtypes

borough       object
town          object
post_code     object
latitude     float64
longitude    float64
dtype: object

### Co-ordinates for London

Getting the geocode for London to help visualize it on the map

In [26]:
london = geocode(address='London, England, GBR')[0]
london_lng_coords = london['location']['x']
london_lat_coords = london['location']['y']
london_lng_coords

-0.1272099999999341

In [27]:
london_lat_coords

51.50642000000005

## Visualize the Map of London

To help visualize the Map of London and the neighbourhoods in London, we make use of the folium package.

In [28]:
# Creating the map of London
map_London = folium.Map(location=[london_lat_coords, london_lng_coords], zoom_start=12)
map_London

# adding markers to map
for latitude, longitude, borough, town in zip(london_merged['latitude'], london_merged['longitude'], london_merged['borough'], london_merged['town']):
    label = '{}, {}'.format(town, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [latitude, longitude],
        radius=5,
        popup=label,
        color='red',
        fill=True
        ).add_to(map_London)  
    
map_London

### Venues in London

To proceed with the next part, we need to define Foursquare API credentials.

Using Foursquare API, we are able to get the venue and venue categories around each neighbourhood in London.

In [40]:
CLIENT_ID = '0Q3QAKIOM540X5ERUIDVVXLICBL2NRZJIP4DQPMDVQ0ZLMA3' 
CLIENT_SECRET = '21O1IJV2ZTMKQMO3NOULGHQEMWC5U3RJ0E2LTA5DNKBY3P4M'
VERSION = '20180605' # Foursquare API version

Defining a function to get the neraby venues in the neighbourhood. This will help us get venue categories which is important for our analysis

In [41]:
LIMIT=100

def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius,
            LIMIT
            )
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighbourhood', 
                  'Neighbourhood Latitude', 
                  'Neighbourhood Longitude', 
                  'Venue', 
                  'Venue Category']
    
    return(nearby_venues)

Getting the venues in London

In [42]:
venues_in_London = getNearbyVenues(london_merged['borough'], london_merged['latitude'], london_merged['longitude'])

Bexley, Greenwich 
Ealing, Hammersmith and Fulham
City
Westminster
Bromley
Islington
Islington
Barnet
Enfield
Wandsworth
Southwark
City
Richmond upon Thames
Barnet
Islington
Wandsworth
Westminster
Bromley
Newham
Ealing
Westminster
Lewisham
Camden
Southwark
Tower Hamlets
Bexley
City
Lewisham
Greenwich
Tower Hamlets
Camden
Haringey
Tower Hamlets
Haringey
Barnet
Brent
Lambeth
Lewisham
Tower Hamlets
Kensington and ChelseaHammersmith and Fulham
Brent
Barnet
Barnet
Southwark
Tower Hamlets
Camden
Tower Hamlets
Waltham Forest
Newham
Islington
Richmond upon Thames
Lewisham
Camden
Westminster
Greenwich
Kensington and Chelsea
Barnet
Westminster
Lewisham
Waltham Forest
Hounslow, Ealing, Hammersmith and Fulham
Brent
Barnet
Lambeth, Wandsworth
Islington
Barnet
Merton
Barnet
Westminster
Barnet, Brent, Camden
Lewisham
Bexley
Haringey
Bromley
Tower Hamlets
Newham
Hackney
Islington
Southwark
Lewisham
Brent
Southwark
Ealing
Kensington and Chelsea
Wandsworth
Southwark
Barnet
Newham
Richmond upon Thames
En

Sampling our data

In [43]:
venues_in_London.head()

Neighbourhood  Neighbourhood Latitude  Neighbourhood Longitude  \
0  Bexley, Greenwich                 51.49245                  0.12127   
1  Bexley, Greenwich                 51.49245                  0.12127   
2  Bexley, Greenwich                 51.49245                  0.12127   
3  Bexley, Greenwich                 51.49245                  0.12127   
4  Bexley, Greenwich                 51.49245                  0.12127   

                              Venue Venue Category  
0                      Lesnes Abbey  Historic Site  
1                       Sainsbury's    Supermarket  
2                              Lidl    Supermarket  
3  Abbey Wood Railway Station (ABW)  Train Station  
4                       Bean @ Work    Coffee Shop

In [44]:
venues_in_London.shape

(10504, 5)

Wow, we have scraped together 10567 records for venues. This will definitely make the clustering interesting.



### Grouping by Venue Categories
We need to now see how many Venue Categories are there for further processing

In [45]:
venues_in_London.groupby('Venue Category').max()

Neighbourhood  \
Venue Category                                                      
Adult Boutique                                          Islington   
African Restaurant                                    Westminster   
American Restaurant                                Waltham Forest   
Antique Shop                                          Westminster   
Arcade                                                Westminster   
...                                                           ...   
Wings Joint                                Hammersmith and Fulham   
Women's Store        Kensington and ChelseaHammersmith and Fulham   
Xinjiang Restaurant                                     Southwark   
Yoga Studio                                           Westminster   
Zoo Exhibit                                                Camden   

                     Neighbourhood Latitude  Neighbourhood Longitude  \
Venue Category                                                         
Adult Boutique                     51.52969                 -0.08697   
African Restaurant                 51.52587                 -0.08808   
American Restaurant                51.63261                  0.02795   
Antique Shop                       51.51651                 -0.11968   
Arcade                             51.51651                 -0.11968   
...                                     ...                      ...   
Wings Joint                        51.54187                 -0.19795   
Women's Store                      51.55457                 -0.11478   
Xinjiang Restaurant                51.47480                 -0.09313   
Yoga Studio                        51.55457                 -0.03558   
Zoo Exhibit                        51.53354                 -0.14606   

                                           Venue  
Venue Category                                    
Adult Boutique       Sh! Women's Erotic Emporium  
African Restaurant            Red Sea Restaurant  
American Restaurant                  Spielburger  
Antique Shop            The London Silver Vaults  
Arcade                        Novelty Automation  
...                                          ...  
Wings Joint                             Wingmans  
Women's Store                 Vivien of Holloway  
Xinjiang Restaurant                    Silk Road  
Yoga Studio                            yogahaven  
Zoo Exhibit                        Penguin Beach  

[294 rows x 4 columns]

We can see 294 records, just goes to show how diverse and interesting the place is.

### One Hot Encoding 
We need to Encode our venue categories to get a better result for our clustering

In [46]:
London_venue_cat = pd.get_dummies(venues_in_London[['Venue Category']], prefix="", prefix_sep="")
London_venue_cat

Adult Boutique  African Restaurant  American Restaurant  Antique Shop  \
0                   0                   0                    0             0   
1                   0                   0                    0             0   
2                   0                   0                    0             0   
3                   0                   0                    0             0   
4                   0                   0                    0             0   
...               ...                 ...                  ...           ...   
10499               0                   0                    0             0   
10500               0                   0                    0             0   
10501               0                   0                    0             0   
10502               0                   0                    0             0   
10503               0                   0                    0             0   

       Arcade  Arepa Restaurant  Argentinian Restaurant  Art Gallery  \
0           0                 0                       0            0   
1           0                 0                       0            0   
2           0                 0                       0            0   
3           0                 0                       0            0   
4           0                 0                       0            0   
...       ...               ...                     ...          ...   
10499       0                 0                       0            0   
10500       0                 0                       0            0   
10501       0                 0                       0            0   
10502       0                 0                       0            0   
10503       0                 0                       0            0   

       Art Museum  Arts & Crafts Store  ...  Vietnamese Restaurant  \
0               0                    0  ...                      0   
1               0                    0  ...                      0   
2               0                    0  ...                      0   
3               0                    0  ...                      0   
4               0                    0  ...                      0   
...           ...                  ...  ...                    ...   
10499           0                    0  ...                      0   
10500           0                    0  ...                      0   
10501           0                    0  ...                      0   
10502           0                    0  ...                      0   
10503           0                    0  ...                      0   

       Warehouse Store  Whisky Bar  Wine Bar  Wine Shop  Wings Joint  \
0                    0           0         0          0            0   
1                    0           0         0          0            0   
2                    0           0         0          0            0   
3                    0           0         0          0            0   
4                    0           0         0          0            0   
...                ...         ...       ...        ...          ...   
10499                0           0         0          0            0   
10500                0           0         0          0            0   
10501                0           0         0          0            0   
10502                0           0         0          0            0   
10503                0           0         0          0            0   

       Women's Store  Xinjiang Restaurant  Yoga Studio  Zoo Exhibit  
0                  0                    0            0            0  
1                  0                    0            0            0  
2                  0                    0            0            0  
3                  0                    0            0            0  
4                  0                    0            0            0  
...              ...                  ...          ... 

Adding Neighbourhood into the mix.

In [47]:
London_venue_cat['Neighbourhood'] = venues_in_London['Neighbourhood'] 

# moving neighborhood column to the first column
fixed_columns = [London_venue_cat.columns[-1]] + list(London_venue_cat.columns[:-1])
London_venue_cat = London_venue_cat[fixed_columns]

London_venue_cat.head()

Neighbourhood  Adult Boutique  African Restaurant  \
0  Bexley, Greenwich                0                   0   
1  Bexley, Greenwich                0                   0   
2  Bexley, Greenwich                0                   0   
3  Bexley, Greenwich                0                   0   
4  Bexley, Greenwich                0                   0   

   American Restaurant  Antique Shop  Arcade  Arepa Restaurant  \
0                    0             0       0                 0   
1                    0             0       0                 0   
2                    0             0       0                 0   
3                    0             0       0                 0   
4                    0             0       0                 0   

   Argentinian Restaurant  Art Gallery  Art Museum  ...  \
0                       0            0           0  ...   
1                       0            0           0  ...   
2                       0            0           0  ...   
3                       0            0           0  ...   
4                       0            0           0  ...   

   Vietnamese Restaurant  Warehouse Store  Whisky Bar  Wine Bar  Wine Shop  \
0                      0                0           0         0          0   
1                      0                0           0         0          0   
2                      0                0           0         0          0   
3                      0                0           0         0          0   
4                      0                0           0         0          0   

   Wings Joint  Women's Store  Xinjiang Restaurant  Yoga Studio  Zoo Exhibit  
0            0              0                    0            0            0  
1            0              0                    0            0            0  
2            0              0                    0            0            0  
3            0              0                    0            0            0  
4            0              0                    0            0            0  

[5 rows x 295 columns]

### Venue categories mean value
We will group the Neighbourhoods and calculate the mean venue categories value in each Neighbourhood

In [48]:
London_grouped = London_venue_cat.groupby('Neighbourhood').mean().reset_index()
London_grouped.head()

Neighbourhood  Adult Boutique  African Restaurant  \
0                 Barnet             0.0                 0.0   
1  Barnet, Brent, Camden             0.0                 0.0   
2                 Bexley             0.0                 0.0   
3      Bexley, Greenwich             0.0                 0.0   
4     Bexley, Greenwich              0.0                 0.0   

   American Restaurant  Antique Shop  Arcade  Arepa Restaurant  \
0             0.007092           0.0     0.0               0.0   
1             0.000000           0.0     0.0               0.0   
2             0.000000           0.0     0.0               0.0   
3             0.000000           0.0     0.0               0.0   
4             0.000000           0.0     0.0               0.0   

   Argentinian Restaurant  Art Gallery  Art Museum  ...  \
0                0.007092          0.0         0.0  ...   
1                0.000000          0.0         0.0  ...   
2                0.000000          0.0         0.0  ...   
3                0.000000          0.0         0.0  ...   
4                0.000000          0.0         0.0  ...   

   Vietnamese Restaurant  Warehouse Store  Whisky Bar  Wine Bar  Wine Shop  \
0               0.007092              0.0         0.0       0.0        0.0   
1               0.000000              0.0         0.0       0.0        0.0   
2               0.000000              0.0         0.0       0.0        0.0   
3               0.000000              0.0         0.0       0.0        0.0   
4               0.000000              0.0         0.0       0.0        0.0   

   Wings Joint  Women's Store  Xinjiang Restaurant  Yoga Studio  Zoo Exhibit  
0          0.0            0.0                  0.0          0.0          0.0  
1          0.0            0.0                  0.0          0.0          0.0  
2          0.0            0.0                  0.0          0.0          0.0  
3          0.0            0.0                  0.0          0.0          0.0  
4          0.0            0.0                  0.0          0.0          0.0  

[5 rows x 295 columns]


Let's make a function to get the top most common venue categories

In [49]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]


There are way too many venue categories, we can take the top 10 to cluster the neighbourhoods.

Creating a function to label the columns of the venue correctly

In [50]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))


### Top venue categories

Getting the top venue categories in London

In [51]:
# create a new dataframe for London
neighborhoods_venues_sorted_london = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted_london['Neighbourhood'] = London_grouped['Neighbourhood']

for ind in np.arange(London_grouped.shape[0]):
    neighborhoods_venues_sorted_london.iloc[ind, 1:] = return_most_common_venues(London_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted_london.head()

Neighbourhood 1st Most Common Venue 2nd Most Common Venue  \
0                 Barnet           Coffee Shop                  Café   
1  Barnet, Brent, Camden           Bus Station  Gym / Fitness Center   
2                 Bexley           Supermarket         Historic Site   
3      Bexley, Greenwich         Historic Site     Convenience Store   
4     Bexley, Greenwich            Supermarket         Historic Site   

  3rd Most Common Venue 4th Most Common Venue       5th Most Common Venue  \
0         Grocery Store    Italian Restaurant                         Pub   
1        Clothing Store           Supermarket                 Zoo Exhibit   
2     Convenience Store           Coffee Shop               Train Station   
3           Sports Club                  Park  Construction & Landscaping   
4         Train Station           Coffee Shop           Convenience Store   

  6th Most Common Venue       7th Most Common Venue 8th Most Common Venue  \
0              Bus Stop                 Supermarket              Pharmacy   
1     Fish & Chips Shop          Falafel Restaurant        Farmers Market   
2           Golf Course  Construction & Landscaping                  Park   
3           Golf Course                    Bus Stop           Flower Shop   
4     Fish & Chips Shop          Falafel Restaurant        Farmers Market   

  9th Most Common Venue 10th Most Common Venue  
0      Sushi Restaurant     Turkish Restaurant  
1  Fast Food Restaurant    Filipino Restaurant  
2              Bus Stop            Fish Market  
3           Flea Market      Food & Drink Shop  
4  Fast Food Restaurant    Filipino Restaurant

## Model Building

### K Means
Let's cluster the city of london to roughly 5 to make it easier to analyze. 

We use the K Means clustering technique to do so.

In [52]:
# set number of clusters
k_num_clusters = 5

London_grouped_clustering = London_grouped.drop('Neighbourhood', 1)

# run k-means clustering
kmeans_london = KMeans(n_clusters=k_num_clusters, random_state=0).fit(London_grouped_clustering)
kmeans_london

KMeans(n_clusters=5, random_state=0)

### Labelling Clustered Data

In [53]:
kmeans_london.labels_

array([1, 0, 2, 2, 2, 0, 1, 0, 1, 0, 1, 1, 1, 1, 1, 1, 3, 1, 1, 1, 1, 1,
       1, 1, 1, 4, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1], dtype=int32)

So our model has labeled the city

In [54]:
neighborhoods_venues_sorted_london.insert(0, 'Cluster Labels', kmeans_london.labels_ +1)

Join London_merged with our neighbourhood venues sorted to add latitude & longitude for each of the neighborhood to prepare it for plotting

In [55]:
london_data = london_merged

london_data = london_data.join(neighborhoods_venues_sorted_london.set_index('Neighbourhood'), on='borough')

london_data.head()

borough    town post_code  latitude  longitude  \
0              Bexley, Greenwich   LONDON       SE2  51.49245    0.12127   
1  Ealing, Hammersmith and Fulham  LONDON    W3, W4  51.51324   -0.26746   
6                            City  LONDON       EC3  51.51200   -0.08058   
7                     Westminster  LONDON       WC2  51.51651   -0.11968   
9                         Bromley  LONDON      SE20  51.41009   -0.05683   

   Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0               3           Supermarket         Historic Site   
1               4         Grocery Store     Indian Restaurant   
6               2           Coffee Shop                 Hotel   
7               2                 Hotel           Coffee Shop   
9               1           Supermarket                 Hotel   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0         Train Station           Coffee Shop     Convenience Store   
1        Breakfast Spot                  Park         Train Station   
6    Italian Restaurant  Gym / Fitness Center                   Pub   
7                  Café        Sandwich Place                   Pub   
9         Grocery Store  Fast Food Restaurant     Convenience Store   

  6th Most Common Venue 7th Most Common Venue       8th Most Common Venue  \
0     Fish & Chips Shop    Falafel Restaurant              Farmers Market   
1           Film Studio               Exhibit          Falafel Restaurant   
6            Restaurant        Sandwich Place                Cocktail Bar   
7    Italian Restaurant               Theater                  Restaurant   
9                  Park     Indian Restaurant  Construction & Landscaping   

  9th Most Common Venue 10th Most Common Venue  
0  Fast Food Restaurant    Filipino Restaurant  
1        Farmers Market   Fast Food Restaurant  
6              Wine Bar     Falafel Restaurant  
7          Burger Joint      French Restaurant  
9             Gastropub                 Bistro


Drop all the NaN values to prevent data skew

In [56]:
london_data_nonan = london_data.dropna(subset=['Cluster Labels'])

### Visualizing the clustered neighbourhood
Let's plot the clusters

In [57]:
map_clusters_london = folium.Map(location=[london_lat_coords, london_lng_coords], zoom_start=12)

# set color scheme for the clusters
x = np.arange(k_num_clusters)
ys = [i + x + (i*x)**2 for i in range(k_num_clusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(london_data_nonan['latitude'], london_data_nonan['longitude'], london_data_nonan['borough'], london_data_nonan['Cluster Labels']):
    label = folium.Popup('Cluster ' + str(int(cluster) +1) + '\n' + str(poi) , parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[int(cluster-1)],
        fill=True,
        fill_color=rainbow[int(cluster-1)]
        ).add_to(map_clusters_london)
        
map_clusters_london

## Examining our Clusters

Cluster 1

In [58]:
london_data_nonan.loc[london_data_nonan['Cluster Labels'] == 1, london_data_nonan.columns[[1] + list(range(5, london_data_nonan.shape[1]))]]

town  Cluster Labels 1st Most Common Venue  \
9               LONDON               1           Supermarket   
29   BECKENHAM, LONDON               1           Supermarket   
61              LONDON               1  Fast Food Restaurant   
69              LONDON               1  Fast Food Restaurant   
100             LONDON               1  Fast Food Restaurant   
121             LONDON               1           Bus Station   
127             LONDON               1           Supermarket   
137             LONDON               1  Fast Food Restaurant   
217             LONDON               1  Fast Food Restaurant   
260             LONDON               1  Fast Food Restaurant   
269             LONDON               1  Fast Food Restaurant   
316             LONDON               1           Supermarket   
319             LONDON               1  Fast Food Restaurant   
357             LONDON               1         Grocery Store   
360             LONDON               1           Supermarket   
377             LONDON               1  Fast Food Restaurant   
442             LONDON               1  Fast Food Restaurant   
453             LONDON               1     Indian Restaurant   
518             LONDON               1  Fast Food Restaurant   

    2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
9                   Hotel         Grocery Store  Fast Food Restaurant   
29                  Hotel         Grocery Store  Fast Food Restaurant   
61          Grocery Store        Sandwich Place              Pharmacy   
69          Grocery Store        Sandwich Place              Pharmacy   
100         Grocery Store        Sandwich Place              Pharmacy   
121  Gym / Fitness Center        Clothing Store           Supermarket   
127                 Hotel         Grocery Store  Fast Food Restaurant   
137         Grocery Store        Sandwich Place              Pharmacy   
217         Grocery Store        Sandwich Place              Pharmacy   
260         Grocery Store        Sandwich Place              Pharmacy   
269         Grocery Store        Sandwich Place              Pharmacy   
316                 Hotel         Grocery Store  Fast Food Restaurant   
319         Grocery Store        Sandwich Place              Pharmacy   
357        Sandwich Place          Liquor Store    Chinese Restaurant   
360                 Hotel         Grocery Store  Fast Food Restaurant   
377         Grocery Store        Sandwich Place              Pharmacy   
442         Grocery Store        Sandwich Place              Pharmacy   
453  Gym / Fitness Center                   Pub              Pharmacy   
518         Grocery Store        Sandwich Place              Pharmacy   

    5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
9       Convenience Store                  Park     Indian Restaurant   
29      Convenience Store                  Park     Indian Restaurant   
61         Discount Store     Indian Restaurant    Chinese Restaurant   
69         Discount Store     Indian Restaurant    Chinese Restaurant   
100        Discount Store     Indian Restaurant    Chinese Restaurant   
121           Zoo Exhibit     Fish & Chips Shop    Falafel Restaurant   
127     Convenience Store                  Park     Indian Restaurant   
137        Discount Store     Indian Restaurant    Chinese Restaurant   
217        Discount Store     Indian Restaurant    Chinese Restaurant   
260        Discount Store     Indian Restaurant    Chinese Restaurant   
269        Discount Store     Indian Restaurant    Chinese Restaurant   
316     Convenience Store                  Park     Indian Restaurant   
319        Discount Store     Indian Restaurant    Chinese Restaurant   
357  Fast Food Restaurant     Convenience Store       Warehouse Store   
360     Convenience Store                  Park     Indian Restaurant   
377        Discount Store     Indian Restaurant    Chinese Restaurant   
442        Discount Store     India

Cluster 2

In [59]:
london_data_nonan.loc[london_data_nonan['Cluster Labels'] == 2, london_data_nonan.columns[[1] + list(range(5, london_data_nonan.shape[1]))]]

town  Cluster Labels 1st Most Common Venue  \
6                    LONDON               2           Coffee Shop   
7                    LONDON               2                 Hotel   
10                   LONDON               2           Coffee Shop   
12                   LONDON               2           Coffee Shop   
14           BARNET, LONDON               2           Coffee Shop   
..                      ...             ...                   ...   
522                  LONDON               2                  Café   
523  LONDON, WOODFORD GREEN               2                 Hotel   
526                  LONDON               2           Coffee Shop   
527                  LONDON               2                   Pub   
529                  LONDON               2                   Pub   

    2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
6                   Hotel    Italian Restaurant  Gym / Fitness Center   
7             Coffee Shop                  Café        Sandwich Place   
10                    Pub                  Café            Food Truck   
12                    Pub                  Café            Food Truck   
14                   Café         Grocery Store    Italian Restaurant   
..                    ...                   ...                   ...   
522                   Pub         Grocery Store           Coffee Shop   
523                  Café                   Pub               Theater   
526                  Café         Grocery Store    Italian Restaurant   
527         Grocery Store              Bus Stop           Coffee Shop   
529                  Café           Coffee Shop         Grocery Store   

     5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
6                      Pub            Restaurant        Sandwich Place   
7                      Pub    Italian Restaurant               Theater   
10   Vietnamese Restaurant  Gym / Fitness Center    Italian Restaurant   
12   Vietnamese Restaurant  Gym / Fitness Center    Italian Restaurant   
14                     Pub              Bus Stop           Supermarket   
..                     ...                   ...                   ...   
522                 Bakery                   Bar         Metro Station   
523    Monument / Landmark                Garden                 Plaza   
526                    Pub              Bus Stop           Supermarket   
527      Indian Restaurant         Historic Site                  Park   
529              Gastropub                 Hotel           Pizza Place   

    8th Most Common Venue 9th Most Common Venue      10th Most Common Venue  
6            Cocktail Bar              Wine Bar          Falafel Restaurant  
7              Restaurant          Burger Joint           French Restaurant  
10                   Park          Cocktail Bar              Breakfast Spot  
12                   Park          Cocktail Bar              Breakfast Spot  
14               Pharmacy      Sushi Restaurant          Turkish Restaurant  
..                    ...                   ...                         ...  
522             BBQ Joint    Seafood Restaurant                        Park  
523                Bakery     French Restaurant                  Steakhouse  
526              Pharmacy      Sushi Restaurant          Turkish Restaurant  
527                  Café        Clothing Store  Construction & Landscaping  
529       Thai Restaurant    Italian Restaurant                        Park  

[281 rows x 12 columns]

Cluster 3

In [60]:
london_data_nonan.loc[london_data_nonan['Cluster Labels'] == 3, london_data_nonan.columns[[1] + list(range(5, london_data_nonan.shape[1]))]]

town  Cluster Labels 1st Most Common Venue  \
0                 LONDON               3           Supermarket   
45   BEXLEYHEATH, LONDON               3           Supermarket   
124               LONDON               3           Supermarket   
167      LONDON, WELLING               3         Historic Site   
291       LONDON, SIDCUP               3           Supermarket   
458        LONDON, ERITH               3         Historic Site   
506               LONDON               3           Supermarket   

    2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0           Historic Site         Train Station           Coffee Shop   
45          Historic Site     Convenience Store           Coffee Shop   
124         Historic Site     Convenience Store           Coffee Shop   
167     Convenience Store           Sports Club                  Park   
291         Historic Site     Convenience Store           Coffee Shop   
458     Convenience Store           Sports Club                  Park   
506         Historic Site     Convenience Store           Coffee Shop   

          5th Most Common Venue 6th Most Common Venue  \
0             Convenience Store     Fish & Chips Shop   
45                Train Station           Golf Course   
124               Train Station           Golf Course   
167  Construction & Landscaping           Golf Course   
291               Train Station           Golf Course   
458  Construction & Landscaping           Golf Course   
506               Train Station           Golf Course   

          7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
0            Falafel Restaurant        Farmers Market  Fast Food Restaurant   
45   Construction & Landscaping                  Park              Bus Stop   
124  Construction & Landscaping                  Park              Bus Stop   
167                    Bus Stop           Flower Shop           Flea Market   
291  Construction & Landscaping                  Park              Bus Stop   
458                    Bus Stop           Flower Shop           Flea Market   
506  Construction & Landscaping                  Park              Bus Stop   

    10th Most Common Venue  
0      Filipino Restaurant  
45             Fish Market  
124            Fish Market  
167      Food & Drink Shop  
291            Fish Market  
458      Food & Drink Shop  
506            Fish Market

Cluster 4

In [61]:
london_data_nonan.loc[london_data_nonan['Cluster Labels'] == 4, london_data_nonan.columns[[1] + list(range(5, london_data_nonan.shape[1]))]]

town  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
1  LONDON               4         Grocery Store     Indian Restaurant   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
1        Breakfast Spot                  Park         Train Station   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
1           Film Studio               Exhibit    Falafel Restaurant   

  9th Most Common Venue 10th Most Common Venue  
1        Farmers Market   Fast Food Restaurant

Cluster 5

In [85]:
london_data_nonan.loc[london_data_nonan['Cluster Labels'] == 5, london_data_nonan.columns[[1] + list(range(5, london_data_nonan.shape[1]))]]

town  Cluster Labels  ... 9th Most Common Venue 10th Most Common Venue
121  LONDON               5  ...    Falafel Restaurant         Farmers Market

[1 rows x 12 columns]



---



# Exploring New York

## Neighbourhoods of New York

### Data Collection

We read the json data with pandas.

In [71]:
import json
!pip install geopy
#!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

In [62]:
!wget -q -O 'newyork_data.json' https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DS0701EN-SkillsNetwork/labs/newyork_data.json
print('Data downloaded!')

Data downloaded!


In [68]:
with open('newyork_data.json') as json_data:
    newyork_data = json.load(json_data)

neighborhoods_data = newyork_data['features']

In [69]:
# define the dataframe columns
column_names = ['Borough', 'Neighborhood', 'Latitude', 'Longitude'] 

# instantiate the dataframe
neighborhoods = pd.DataFrame(columns=column_names)

### Data Preprocessing 

We break down each of the nested fields and create the dataframe that we need

In [70]:
for data in neighborhoods_data:
    borough = neighborhood_name = data['properties']['borough'] 
    neighborhood_name = data['properties']['name']
        
    neighborhood_latlon = data['geometry']['coordinates']
    neighborhood_lat = neighborhood_latlon[1]
    neighborhood_lon = neighborhood_latlon[0]
    
    neighborhoods = neighborhoods.append({'Borough': borough,
                                          'Neighborhood': neighborhood_name,
                                          'Latitude': neighborhood_lat,
                                          'Longitude': neighborhood_lon}, ignore_index=True)

neighborhoods.head()

Borough Neighborhood   Latitude  Longitude
0   Bronx    Wakefield  40.894705 -73.847201
1   Bronx   Co-op City  40.874294 -73.829939
2   Bronx  Eastchester  40.887556 -73.827806
3   Bronx    Fieldston  40.895437 -73.905643
4   Bronx    Riverdale  40.890834 -73.912585

#### Use geopy library to get the latitude and longitude values of New York City.

In order to define an instance of the geocoder, we need to define a user_agent. We will name our agent <em>ny_explorer</em>, as shown below.

In [72]:
address = 'New York City, NY'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of New York City are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of New York City are 40.7127281, -74.0060152.


We have managed to collect 1300 records.

## Visualize the Map of New York

In [73]:
# create map of New York using latitude and longitude values
map_newyork = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(neighborhoods['Latitude'], neighborhoods['Longitude'], neighborhoods['Borough'], neighborhoods['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_newyork)  
    
map_newyork

In [74]:
manhattan_data = neighborhoods[neighborhoods['Borough'] == 'Manhattan'].reset_index(drop=True)
manhattan_data.head()

Borough        Neighborhood   Latitude  Longitude
0  Manhattan         Marble Hill  40.876551 -73.910660
1  Manhattan           Chinatown  40.715618 -73.994279
2  Manhattan  Washington Heights  40.851903 -73.936900
3  Manhattan              Inwood  40.867684 -73.921210
4  Manhattan    Hamilton Heights  40.823604 -73.949688

In [75]:
address = 'Manhattan, NY'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Manhattan are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Manhattan are 40.7896239, -73.9598939.


In [76]:
# create map of Manhattan using latitude and longitude values
map_manhattan = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, label in zip(manhattan_data['Latitude'], manhattan_data['Longitude'], manhattan_data['Neighborhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_manhattan)  
    
map_manhattan

### Venues in Manhattan

Using our previously defined function. Let's get the neaby venues present in each neighbourhood of Manhattan

In [77]:
manhattan_venues = getNearbyVenues(names=manhattan_data['Neighborhood'],
                                   latitudes=manhattan_data['Latitude'],
                                   longitudes=manhattan_data['Longitude']
                                  )

Marble Hill
Chinatown
Washington Heights
Inwood
Hamilton Heights
Manhattanville
Central Harlem
East Harlem
Upper East Side
Yorkville
Lenox Hill
Roosevelt Island
Upper West Side
Lincoln Square
Clinton
Midtown
Murray Hill
Chelsea
Greenwich Village
East Village
Lower East Side
Tribeca
Little Italy
Soho
West Village
Manhattan Valley
Morningside Heights
Gramercy
Battery Park City
Financial District
Carnegie Hill
Noho
Civic Center
Midtown South
Sutton Place
Turtle Bay
Tudor City
Stuyvesant Town
Flatiron
Hudson Yards


Sampling our Data

In [78]:
manhattan_venues.head()

Neighbourhood  Neighbourhood Latitude  Neighbourhood Longitude  \
0   Marble Hill               40.876551                -73.91066   
1   Marble Hill               40.876551                -73.91066   
2   Marble Hill               40.876551                -73.91066   
3   Marble Hill               40.876551                -73.91066   
4   Marble Hill               40.876551                -73.91066   

           Venue Venue Category  
0       Arturo's    Pizza Place  
1    Bikram Yoga    Yoga Studio  
2  Tibbett Diner          Diner  
3      Starbucks    Coffee Shop  
4        Dunkin'     Donut Shop

In [79]:
print(manhattan_venues.shape)

(3205, 5)


We have managed to collect 3205 venue records for the neighbourhoods in Manhattan

### Grouping by Venue Categories
We need to now see how many Venue Categories are there for further processing

In [80]:
manhattan_venues.groupby('Venue Category').max()

Neighbourhood  Neighbourhood Latitude  \
Venue Category                                                    
Accessories Store          West Village               40.851903   
Adult Boutique             Sutton Place               40.760280   
African Restaurant       Central Harlem               40.815976   
American Restaurant        West Village               40.867684   
Antique Shop               Civic Center               40.715229   
...                                 ...                     ...   
Wine Bar                      Yorkville               40.867684   
Wine Shop                     Yorkville               40.867684   
Wings Joint            Manhattan Valley               40.797307   
Women's Store        Washington Heights               40.851903   
Yoga Studio             Upper West Side               40.876551   

                     Neighbourhood Longitude                            Venue  
Venue Category                                                                 
Accessories Store                 -73.936900              Venus by Maria Tash  
Adult Boutique                    -73.963556  The Pleasure Chest New York UES  
African Restaurant                -73.943211              Ponty Bistro Harlem  
American Restaurant               -73.921210                        l'estudio  
Antique Shop                      -74.005415          Philip Williams Posters  
...                                      ...                              ...  
Wine Bar                          -73.921210                             bibi  
Wine Shop                         -73.921210                          convive  
Wings Joint                       -73.964286                       Koko Wings  
Women's Store                     -73.936900               New York & Company  
Yoga Studio                       -73.910660                     mang'Oh yoga  

[321 rows x 4 columns]

321 records in Manhattan. Manhattan is also multicultural and diverse as London

### One Hot Encoding 
We need to Encode our venue categories to get a better result for our clustering

In [82]:
# one hot encoding
manhattan_onehot = pd.get_dummies(manhattan_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
manhattan_onehot['Neighbourhood'] = manhattan_venues['Neighbourhood'] 

# move neighborhood column to the first column
fixed_columns = [manhattan_onehot.columns[-1]] + list(manhattan_onehot.columns[:-1])
manhattan_onehot = manhattan_onehot[fixed_columns]

manhattan_onehot.head()

Neighbourhood  Accessories Store  Adult Boutique  African Restaurant  \
0   Marble Hill                  0               0                   0   
1   Marble Hill                  0               0                   0   
2   Marble Hill                  0               0                   0   
3   Marble Hill                  0               0                   0   
4   Marble Hill                  0               0                   0   

   American Restaurant  Antique Shop  Arepa Restaurant  \
0                    0             0                 0   
1                    0             0                 0   
2                    0             0                 0   
3                    0             0                 0   
4                    0             0                 0   

   Argentinian Restaurant  Art Gallery  Art Museum  ...  Video Store  \
0                       0            0           0  ...            0   
1                       0            0           0  ...            0   
2                       0            0           0  ...            0   
3                       0            0           0  ...            0   
4                       0            0           0  ...            0   

   Vietnamese Restaurant  Volleyball Court  Waterfront  Whisky Bar  Wine Bar  \
0                      0                 0           0           0         0   
1                      0                 0           0           0         0   
2                      0                 0           0           0         0   
3                      0                 0           0           0         0   
4                      0                 0           0           0         0   

   Wine Shop  Wings Joint  Women's Store  Yoga Studio  
0          0            0              0            0  
1          0            0              0            1  
2          0            0              0            0  
3          0            0              0            0  
4          0            0              0            0  

[5 rows x 322 columns]

### Venue categories mean value
We will group the Neighbourhoods and calculate the mean venue categories value in each Neighbourhood

In [83]:
manhattan_grouped = manhattan_onehot.groupby('Neighbourhood').mean().reset_index()
manhattan_grouped

Neighbourhood  Accessories Store  Adult Boutique  \
0     Battery Park City           0.000000            0.00   
1         Carnegie Hill           0.000000            0.00   
2        Central Harlem           0.000000            0.00   
3               Chelsea           0.000000            0.00   
4             Chinatown           0.000000            0.00   
5          Civic Center           0.000000            0.00   
6               Clinton           0.000000            0.00   
7           East Harlem           0.000000            0.00   
8          East Village           0.000000            0.00   
9    Financial District           0.000000            0.00   
10             Flatiron           0.000000            0.00   
11             Gramercy           0.000000            0.00   
12    Greenwich Village           0.010000            0.00   
13     Hamilton Heights           0.000000            0.00   
14         Hudson Yards           0.000000            0.00   
15               Inwood           0.000000            0.00   
16           Lenox Hill           0.000000            0.00   
17       Lincoln Square           0.000000            0.00   
18         Little Italy           0.000000            0.00   
19      Lower East Side           0.000000            0.00   
20     Manhattan Valley           0.000000            0.00   
21       Manhattanville           0.000000            0.00   
22          Marble Hill           0.000000            0.00   
23              Midtown           0.000000            0.00   
24        Midtown South           0.000000            0.00   
25  Morningside Heights           0.000000            0.00   
26          Murray Hill           0.000000            0.00   
27                 Noho           0.000000            0.00   
28     Roosevelt Island           0.000000            0.00   
29                 Soho           0.000000            0.00   
30      Stuyvesant Town           0.000000            0.00   
31         Sutton Place           0.000000            0.01   
32              Tribeca           0.000000            0.00   
33           Tudor City           0.000000            0.00   
34           Turtle Bay           0.000000            0.00   
35      Upper East Side           0.000000            0.00   
36      Upper West Side           0.010204            0.00   
37   Washington Heights           0.011765            0.00   
38         West Village           0.010000            0.00   
39            Yorkville           0.000000            0.00   

    African Restaurant  American Restaurant  Antique Shop  Arepa Restaurant  \
0             0.000000             0.000000          0.00          0.000000   
1             0.000000             0.000000          0.00          0.000000   
2             0.066667             0.044444          0.00          0.000000   
3             0.000000             0.040000          0.00          0.010000   
4             0.000000             0.040000          0.00          0.000000   
5             0.000000             0.030000          0.01          0.000000   
6             0.000000             0.050000          0.00          0.000000   
7             0.000000             0.000000          0.00          0.000000   
8             0.000000             0.010000          0.00          0.000000   
9             0.000000             0.040000          0.00          0.000000   
10            0.000000             0.030000          0.00          0.000000   
11            0.000000             0.042553          0.00          0.000000   
12            0.000000             0.030000          0.00          0.000000   
13            0.000000             0.000000          0.00          0.000000   
14            0.000000             0.048387          0.00          0.000000   
15            0.000000             0.035088          0.00          0.000000   
16            0.000000             0.000000          0.00          0.000000   
17            0.000000             0.021739  

### Top venue categories
 
Reusing our previously defined function to get the top venue categories in the neighbourhoods of Manhattan.

In [84]:
# create a new dataframe for Manhattan
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighbourhood'] = manhattan_grouped['Neighbourhood']

for ind in np.arange(manhattan_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(manhattan_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighbourhood 1st Most Common Venue 2nd Most Common Venue  \
0  Battery Park City                  Park                 Hotel   
1      Carnegie Hill           Coffee Shop                  Café   
2     Central Harlem    African Restaurant    Chinese Restaurant   
3            Chelsea           Coffee Shop           Art Gallery   
4          Chinatown    Chinese Restaurant                Bakery   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0                   Gym           Coffee Shop         Memorial Site   
1             Bookstore    Italian Restaurant  Gym / Fitness Center   
2                   Bar    Seafood Restaurant   American Restaurant   
3                Bakery   American Restaurant        Ice Cream Shop   
4          Cocktail Bar   American Restaurant          Noodle House   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0         Shopping Mall                 Plaza          Burger Joint   
1                   Gym     French Restaurant           Yoga Studio   
2     French Restaurant        Cosmetics Shop   Fried Chicken Joint   
3    Italian Restaurant                  Park             Bookstore   
4    Salon / Barbershop   Shanghai Restaurant     Hotpot Restaurant   

  9th Most Common Venue 10th Most Common Venue  
0          Gourmet Shop             Playground  
1             Wine Shop  Vietnamese Restaurant  
2  Caribbean Restaurant                   Café  
3          Cycle Studio           Cupcake Shop  
4                   Spa           Dessert Shop

## Model Building

### K Means
Let's cluster the city of Manhattan to roughly 5 to make it easier to analyze. 

We use the K Means clustering technique to do so.

In [86]:
# set number of clusters
kclusters = 5

manhattan_grouped_clustering = manhattan_grouped.drop('Neighbourhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(manhattan_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([1, 1, 1, 1, 1, 1, 1, 3, 0, 1], dtype=int32)

### Labelling Clustered Data

In [87]:
kmeans.labels_

array([1, 1, 1, 1, 1, 1, 1, 3, 0, 1, 1, 1, 0, 0, 1, 3, 0, 0, 0, 1, 1, 3,
       4, 1, 1, 3, 1, 0, 1, 0, 2, 1, 0, 3, 1, 1, 0, 0, 0, 0], dtype=int32)

So our model has labeled the city, we insert it in our data.

In [88]:
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

Join *paris_combined_data* with our neighbourhood venues sorted to add latitude & longitude for each of the neighborhood to prepare it for plotting

In [90]:
manhattan_merged = manhattan_data

# merge manhattan_grouped with manhattan_data to add latitude/longitude for each neighborhood
manhattan_merged = manhattan_merged.join(neighborhoods_venues_sorted.set_index('Neighbourhood'), on='Neighbourhood')

manhattan_merged.head() # check the last columns!

Borough        Neighborhood   Latitude  Longitude  Cluster Labels  \
0  Manhattan         Marble Hill  40.876551 -73.910660               4   
1  Manhattan           Chinatown  40.715618 -73.994279               1   
2  Manhattan  Washington Heights  40.851903 -73.936900               0   
3  Manhattan              Inwood  40.867684 -73.921210               3   
4  Manhattan    Hamilton Heights  40.823604 -73.949688               0   

  1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
0           Coffee Shop        Sandwich Place        Discount Store   
1    Chinese Restaurant                Bakery          Cocktail Bar   
2                  Café                Bakery         Grocery Store   
3    Mexican Restaurant                  Café                Lounge   
4           Pizza Place           Coffee Shop                  Café   

  4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
0                   Gym       Supplement Shop            Donut Shop   
1   American Restaurant          Noodle House    Salon / Barbershop   
2     Mobile Phone Shop                  Bank        Sandwich Place   
3            Restaurant                  Park    Chinese Restaurant   
4    Mexican Restaurant          Cocktail Bar     Indian Restaurant   

  7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
0        Tennis Stadium            Kids Store              Pharmacy   
1   Shanghai Restaurant     Hotpot Restaurant                   Spa   
2           Coffee Shop                  Park    Spanish Restaurant   
3                Bakery    Frozen Yogurt Shop  Caribbean Restaurant   
4          Liquor Store      Sushi Restaurant                  Park   

  10th Most Common Venue  
0            Yoga Studio  
1           Dessert Shop  
2          Deli / Bodega  
3    American Restaurant  
4          Deli / Bodega


Drop all the NaN values to prevent data skew

In [91]:
manhattan_merged = manhattan_merged.dropna(subset=['Cluster Labels'])

### Visualizing the clustered neighbourhood
Let's plot the clusters

In [92]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(manhattan_merged['Latitude'], manhattan_merged['Longitude'], manhattan_merged['Neighborhood'], manhattan_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

## Examining our Clusters

Cluster 1

In [93]:
manhattan_merged.loc[manhattan_merged['Cluster Labels'] == 0, manhattan_merged.columns[[1] + list(range(5, manhattan_merged.shape[1]))]]

Neighborhood 1st Most Common Venue  2nd Most Common Venue  \
2   Washington Heights                  Café                 Bakery   
4     Hamilton Heights           Pizza Place            Coffee Shop   
9            Yorkville    Italian Restaurant            Coffee Shop   
10          Lenox Hill    Italian Restaurant       Sushi Restaurant   
12     Upper West Side    Italian Restaurant                    Bar   
13      Lincoln Square                 Plaza  Performing Arts Venue   
18   Greenwich Village    Italian Restaurant       Sushi Restaurant   
19        East Village                   Bar            Pizza Place   
21             Tribeca                  Park    American Restaurant   
22        Little Italy                Bakery                   Café   
23                Soho        Clothing Store     Italian Restaurant   
24        West Village    Italian Restaurant    American Restaurant   
31                Noho    Italian Restaurant            Pizza Place   

      3rd Most Common Venue 4th Most Common Venue     5th Most Common Venue  \
2             Grocery Store     Mobile Phone Shop                      Bank   
4                      Café    Mexican Restaurant              Cocktail Bar   
9                       Gym                   Bar             Deli / Bodega   
10              Coffee Shop           Pizza Place              Cocktail Bar   
12                     Café     Indian Restaurant               Coffee Shop   
13             Concert Hall    Italian Restaurant                      Café   
18           Clothing Store                  Café         Indian Restaurant   
19       Mexican Restaurant              Wine Bar               Coffee Shop   
21                 Wine Bar    Italian Restaurant               Coffee Shop   
22       Italian Restaurant       Bubble Tea Shop        Chinese Restaurant   
23              Coffee Shop              Boutique  Mediterranean Restaurant   
24  New American Restaurant           Coffee Shop                      Park   
31              Art Gallery    Mexican Restaurant                     Hotel   

       6th Most Common Venue          7th Most Common Venue  \
2             Sandwich Place                    Coffee Shop   
4          Indian Restaurant                   Liquor Store   
9           Sushi Restaurant            Japanese Restaurant   
10             Deli / Bodega                            Gym   
12                  Wine Bar                    Pizza Place   
13                   Theater                         Bakery   
18       American Restaurant                            Gym   
19            Ice Cream Shop  Vegetarian / Vegan Restaurant   
21                       Spa                           Café   
22  Mediterranean Restaurant                 Ice Cream Shop   
23                Shoe Store                         Bakery   
24              Cocktail Bar                       Wine Bar   
31               Yoga Studio                    Coffee Shop   

   8th Most Common Venue 9th Most Common Venue    10th Most Common Venue  
2                   Park    Spanish Restaurant             Deli / Bodega  
4       Sushi Restaurant                  Park             Deli / Bodega  
9              Wine Shop    Mexican Restaurant                     Diner  
10  Gym / Fitness Center                  Café              Burger Joint  
12                Bakery        Ice Cream Shop  Mediterranean Restaurant  
13     French Restaurant   Indie Movie Theater                 Wine Shop  
18              Boutique       Bubble Tea Shop        Chinese Restaurant  
19    Italian Restaurant          Cocktail Bar                 Speakeasy  
21      Greek Restaurant     French Restaurant                Skate Park  
22          Cocktail Bar           Pizza Place               Coffee Shop  
23                  Café     French Restaurant               Pizza Place  
24             Jazz Club               Theater               Pizza Place  
31        Sandwich Place              Wine Bar    

Cluster 2

In [104]:
manhattan_merged.loc[manhattan_merged['Cluster Labels'] == 1, manhattan_merged.columns[[1] + list(range(5, manhattan_merged.shape[1]))]]

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
1            Chinatown    Chinese Restaurant                Bakery   
6       Central Harlem    African Restaurant    Chinese Restaurant   
8      Upper East Side               Exhibit    Italian Restaurant   
11    Roosevelt Island         Deli / Bodega   Japanese Restaurant   
14             Clinton               Theater  Gym / Fitness Center   
15             Midtown                 Hotel                Bakery   
16         Murray Hill   Japanese Restaurant           Coffee Shop   
17             Chelsea           Coffee Shop           Art Gallery   
20     Lower East Side    Chinese Restaurant              Pharmacy   
25    Manhattan Valley                   Bar           Coffee Shop   
27            Gramercy                   Bar    Italian Restaurant   
28   Battery Park City                  Park                 Hotel   
29  Financial District           Coffee Shop           Pizza Place   
30       Carnegie Hill           Coffee Shop                  Café   
32        Civic Center           Coffee Shop                   Spa   
33       Midtown South     Korean Restaurant                 Hotel   
34        Sutton Place    Italian Restaurant           Coffee Shop   
35          Turtle Bay    Italian Restaurant           Coffee Shop   
38            Flatiron    Italian Restaurant   Japanese Restaurant   
39        Hudson Yards  Gym / Fitness Center                 Hotel   

      3rd Most Common Venue     4th Most Common Venue   5th Most Common Venue  \
1              Cocktail Bar       American Restaurant            Noodle House   
6                       Bar        Seafood Restaurant     American Restaurant   
8               Coffee Shop                    Bakery    Gym / Fitness Center   
11    Outdoors & Recreation          Greek Restaurant             Supermarket   
14      American Restaurant            Sandwich Place             Coffee Shop   
15              Coffee Shop                Steakhouse                 Theater   
16                    Hotel      Gym / Fitness Center          Sandwich Place   
17                   Bakery       American Restaurant          Ice Cream Shop   
20              Coffee Shop                      Café                  Bakery   
25              Yoga Studio               Pizza Place              Playground   
27      American Restaurant               Coffee Shop         Thai Restaurant   
28                      Gym               Coffee Shop           Memorial Site   
29             Cocktail Bar                       Bar     American Restaurant   
30                Bookstore        Italian Restaurant    Gym / Fitness Center   
32     Gym / Fitness Center               Yoga Studio       French Restaurant   
33      Japanese Restaurant            Cosmetics Shop            Burger Joint   
34                     Park      Gym / Fitness Center  Furniture / Home Store   
35         Sushi Restaurant       Japanese Restaurant                    Park   
38  New American Restaurant  Mediterranean Restaurant            Cycle Studio   
39       Italian Restaurant       American Restaurant             Coffee Shop   

   6th Most Common Venue 7th Most Common Venue   8th Most Common Venue  \
1     Salon / Barbershop   Shanghai Restaurant       Hotpot Restaurant   
6      French Restaurant        Cosmetics Shop     Fried Chicken Joint   
8            Yoga Studio        Cosmetics Shop       French Restaurant   
11       Bubble Tea Shop     Food & Drink Shop            Soccer Field   
14                   Gym                   Spa                   Hotel   
15   Sporting Goods Shop        Clothing Store          Sandwich Place   
16   American Restaurant                   Bar              Restaurant   
17    Italian Restaurant                  Park               Bookstore   
20   Japanese Restaurant           Art Gallery             Pizza Place   
25       Thai Restaurant    Mexican Restaurant          Cosmetics Shop   
27            Bagel Sho

Cluster 3

In [103]:
manhattan_merged.loc[manhattan_merged['Cluster Labels'] == 2, manhattan_merged.columns[[1] + list(range(5, manhattan_merged.shape[1]))]]

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
37  Stuyvesant Town                  Park           Pet Service   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
37          Cocktail Bar       Harbor / Marina  Gym / Fitness Center   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
37        Baseball Field                   Bar                Bistro   

   9th Most Common Venue 10th Most Common Venue  
37              Heliport         Farmers Market

Cluster 4

In [100]:
manhattan_merged.loc[manhattan_merged['Cluster Labels'] == 3, manhattan_merged.columns[[1] + list(range(5, manhattan_merged.shape[1]))]]

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
3                Inwood    Mexican Restaurant                  Café   
5        Manhattanville    Seafood Restaurant           Coffee Shop   
7           East Harlem    Mexican Restaurant                Bakery   
26  Morningside Heights                  Park   American Restaurant   
36           Tudor City                  Café                  Park   

   3rd Most Common Venue 4th Most Common Venue      5th Most Common Venue  \
3                 Lounge            Restaurant                       Park   
5          Deli / Bodega    Italian Restaurant         Mexican Restaurant   
7        Thai Restaurant         Deli / Bodega  Latin American Restaurant   
26           Coffee Shop             Bookstore               Burger Joint   
36    Mexican Restaurant      Asian Restaurant                      Diner   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
3     Chinese Restaurant                Bakery    Frozen Yogurt Shop   
5            Bus Station                Lounge              Boutique   
7         Sandwich Place                   Spa          Liquor Store   
26                  Café         Deli / Bodega                   Pub   
36         Deli / Bodega           Coffee Shop            Restaurant   

            9th Most Common Venue 10th Most Common Venue  
3            Caribbean Restaurant    American Restaurant  
5                Sushi Restaurant            Supermarket  
7                      Taco Place            Gas Station  
26  Paper / Office Supplies Store     Seafood Restaurant  
36                         Garden       Sushi Restaurant

Cluster 5

In [101]:
manhattan_merged.loc[manhattan_merged['Cluster Labels'] == 4, manhattan_merged.columns[[1] + list(range(5, manhattan_merged.shape[1]))]]

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
0  Marble Hill           Coffee Shop        Sandwich Place   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0        Discount Store                   Gym       Supplement Shop   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0            Donut Shop        Tennis Stadium            Kids Store   

  9th Most Common Venue 10th Most Common Venue  
0              Pharmacy            Yoga Studio

# Results and Discussion


The neighbourhoods of London are very mulitcultural. There are a lot of different cusines including Indian, Italian, Turkish and Chinese. London seems to take a step further in this direction by having a lot of Restaurants, bars, juice bars, coffee shops, Fish and Chips shop and Breakfast spots. It has a lot of shopping options too with that of the Flea markets, flower shops, fish markets, Fishing stores, clothing stores. The main modes of transport seem to be Buses and trains. For leisure, the neighbourhoods are set up to have lots of parks, golf courses, zoo, gyms and Historic sites. Overall, the city of London offers a multicultural, diverse and certainly an entertaining experience.

New York is relatively small in size geographically. It has a wide variety of cusines and eateries including Italian, Thai, Cambodian, Indian, Chinese etc. There are a lot of hangout spots including many Restaurants and Bars. Different means of public transport in New York which includes buses, bikes, boats or ferries. For leisure and sight seeing, there are a lot of Plazas, Trails, Parks, Historic sites, clothing shops, Art galleries and Museums. Overall, New York seems like the relaxing vacation spot with a mix of lakes, historic spots and a wide variety of cusines to try out.


# Conclusion

The purpose of this project was to explore the cities of London and New York and see how attractive it is to potential tourists and migrants. We explored both the cities based on their postal codes and then extrapolated the common venues present in each of the neighbourhoods finally concluding with clustering similar neighbourhoods together.

We could see that each of the neighbourhoods in both the cities have a wide variety of experiences to offer which is unique in it's own way. The cultural diversity is quite evident which also gives the feeling of a sense of inclusion.

Both new York and London seem to offer a vacation stay or a romantic getaway with a lot of places to explore, beautiful landscapes, amazing food and a wide variety of culture. Overall, it's upto the stakeholders to decide which experience they would prefer more and which would more to their liking.